src/backend/parser/parse_param.c

```diff
@@ -115,7 +117,14 @@ fixed_paramref_hook(ParseState *pstate, ParamRef *pref)
 	param->paramid = paramno;
 	param->paramtype = parstate->paramTypes[paramno - 1];
 	param->paramtypmod = -1;
-	param->paramcollid = get_typcollation(param->paramtype);
+	if (handle_param_collation_hook)
+	{
+		param->paramcollid = handle_param_collation_hook(param);
+	}
+	else
+	{
+		param->paramcollid = get_typcollation(param->paramtype);
+	}
```

Comment:
Why not rewrite fixed_paramref_hook in TSQL extension and use pre_parse_analyze_hook to replace with
```
pstate->p_paramref_hook = tsql_fixed_paramref_hook
```
Also, I noticed that get_typcollation already has the same check as set_param_collation. So, I'm not sure why we need this change.

Reply:
This special handling is required for PG dialect while checking for constraints, eg., foreign key (FK) constraint. If we try to insert some values on T-SQL table then PG in turn will create query like `SELECT 1 FROM ONLY \"master_dbo\".\"t5\" x WHERE \"a\" OPERATOR(pg_catalog.=) $1 FOR KEY SHARE OF x` to evaluate FK which would be done in PG dialect. 

If we set data type default collation (which will always be same as `babelfishpg_tsql.server_collation_name`) then we may face collation conflict error is T-SQL table is defined under non-default database level collation (ie. `db collation` is not equal `babelfishpg_tsql.server_collation_name`).

This special handling will set `DEFAULT_COLLATION_OID` to param so that other collation (collation of left operand) will be picked up always. Note - collation of left operand will always be defined in all cases.